File: packages/core/src/axes/axis.ts
import type { Scale } from "../scales/scale.js";
import type { AxisConfig, Tick } from "../types.js";
import { resolveTicks, withMinorTicks } from "./ticks.js";

/**
 * Owns the tick configuration for one axis and resolves the final tick list
 * against a scale. Modes: auto (delegates to the scale), custom (array), or
 * generator (function of the range) — plus major/minor, per-tick labels, and
 * `addTicks` layering.
 */
export class Axis {
  config: AxisConfig;

  constructor(config: AxisConfig = {}) {
    this.config = config;
  }

  update(patch: Partial<AxisConfig>): void {
    this.config = { ...this.config, ...patch };
  }

  /** Resolve the concrete tick list (labels filled) for the scale's domain. */
  resolve(scale: Scale): Tick[] {
    const [min, max] = scale.domain;
    const explicit = resolveTicks(this.config.ticks, min, max);

    let ticks: Tick[];
    if (explicit) {
      ticks = explicit;
    } else {
      let major = scale.ticks();
      // Auto minor ticks only make sense for linear-style scales; log/time
      // scales already emit their own minor ticks.
      const minor = this.config.minorTicks;
      if (minor && scale.type === "linear") {
        major = withMinorTicks(major, minor === true ? 4 : minor);
      }
      if (this.config.addTicks?.length) {
        const extra = this.config.addTicks.map((e) =>
          typeof e === "number" ? { value: e } : e,
        );
        major = [...major, ...extra].sort((a, b) => a.value - b.value);
      }
      ticks = major;
    }

    const fmt = this.config.format ?? ((v: number) => scale.formatTick(v));
    return ticks
      .filter((t) => t.value >= min && t.value <= max)
      .map((t) => ({
        value: t.value,
        label: t.minor ? "" : (t.label ?? fmt(t.value)),
        minor: t.minor ?? false,
        grid: t.grid ?? !t.minor,
      }));
  }
}

File: native/src/axes/axis.cpp
#include "axes/axis.hpp"

#include "scale.hpp"

namespace photon {

void Axis::set_explicit_ticks(std::vector<Tick> ticks) {
  explicit_ = std::move(ticks);
  dirty_ = true;
}

void Axis::set_config(render::AxisConfig new_config) {
  config = std::move(new_config);
  dirty_ = true;
}

const std::vector<Tick>& Axis::resolve(const Scale& scale) {
  if (!dirty_ && cached_lo_ == scale.lo && cached_hi_ == scale.hi && cached_type_ == scale.type) {
    return resolved_;
  }

  std::vector<Tick> source;
  if (!explicit_.empty()) {
    source = explicit_;
  } else {
    source = scale.ticks();
    // Auto minor ticks only make sense on a linear axis; log and time scales
    // already emit their own, which is why the web core gates it the same way.
    if (config.minor_ticks > 0 && scale.type == PH_SCALE_LINEAR) {
      source = with_minor_ticks(source, config.minor_ticks);
    }
  }

  resolved_.clear();
  resolved_.reserve(source.size());
  for (const Tick& tick : source) {
    // Ticks outside the visible domain are dropped rather than clamped, so a
    // pan does not pile labels up against the edge.
    if (tick.value < scale.lo || tick.value > scale.hi) continue;
    Tick out;
    out.value = tick.value;
    out.minor = tick.minor;
    out.grid = tick.grid;
    out.label = tick.minor ? std::string() : (tick.label.empty() ? scale.format_tick(tick.value)
                                                                : tick.label);
    resolved_.push_back(std::move(out));
  }

  dirty_ = false;
  cached_lo_ = scale.lo;
  cached_hi_ = scale.hi;
  cached_type_ = scale.type;
  return resolved_;
}

}  // namespace photon
